### PR TITLE
Preferred facility

### PR DESCRIPTION
**Pull request by lihanli**
## Description of change
add preferred facility api and database tables

## Original issue(s)
https://github.com/department-of-veterans-affairs/va.gov-team/issues/16107

## Things to know about this PR

swagger docs updated

### CODE REVIEW

**Review by thilton-oddball — CHANGES_REQUESTED** [review]
We prefer to have individual migrations broken out into their own PRs as opposed to including them with other code changes.

**Comment by lihanli** [comment]
@thilton-oddball thanks for the info, the migrations have been moved out of this PR

File: app/controllers/v0/preferred_facilities_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        render(json: preferred_facility)
+      else
+        raise Common::Exceptions::ValidationErrors, preferred_facility
+      end
+    end
+
+    private
+
+    def current_user_preferred_facilities
+      current_user.account.preferred_facilities
```
Comment (kfrz) [inline app/controllers/v0/preferred_facilities_controller.rb:34]: What should happen if the user doesn't have an account

File: app/models/preferred_facility.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class PreferredFacility < ApplicationRecord
```
Comment (kfrz) [inline app/models/preferred_facility.rb:3]: Please add class-level YARD documentation

File: spec/models/preferred_facility_spec.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    it 'validates presence of user' do
+      expect_attr_invalid(preferred_facility, :user, "can't be blank")
+    end
+
+    it 'validates facility_code in user list' do
+      facility = build(:preferred_facility, facility_code: '111')
+      expect_attr_invalid(facility, :facility_code, "must be included in user's va treatment facilities list")
+    end
+
+    context 'uniqueness of facility_code' do
```
Comment (kfrz) [inline spec/models/preferred_facility_spec.rb:26]: Does this need a context block? Feels awkward here to read.

**Review by kfrz — DISMISSED** [review]
Minor change request re: documentation, otherwise this looks great! Nice specs.

File: app/controllers/v0/preferred_facilities_controller.rb (hunk 1 above)
Comment (lihanli) [reply app/controllers/v0/preferred_facilities_controller.rb:34]: https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/models/user.rb#L48
an account model will be created if it doesn't already exist for a user

**Comment by lihanli** [comment]
@kfrz thanks for the review, i added the changes

**Review by thilton-oddball — APPROVED** [review]
This all looks good!

**Review by kfrz — APPROVED** [review]

**Review by ericboehs — APPROVED** [review]

File: spec/controllers/v0/preferred_facilities_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
```
Comment (bosawt) [inline spec/controllers/v0/preferred_facilities_controller_spec.rb:3]: These specs are great!

**Review by bosawt — COMMENTED** [review]
Nice work! I only had a comment about moving the `destroy` logic out of the render and into a private function, but otherwise looks good!

**Comment by lihanli** [comment]
@bosawt thanks for the review, i added the changes

File: app/controllers/v0/preferred_facilities_controller.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  class PreferredFacilitiesController < ApplicationController
+    def index
+      render(
+        json: current_user_preferred_facilities
+      )
+    end
+
+    def destroy
+      render(
+        json: current_user_preferred_facilities.find(params[:id]).destroy!
```
[the remark this answers is not in the document]
Comment (bosawt) [reply app/controllers/v0/preferred_facilities_controller.rb]: > @bosawt thanks for the review, i added the changes

Looks good!

**Review by bosawt — APPROVED** [review]

**Pull request merged**